crates/red_knot_python_semantic/src/types.rs

```diff
@@ -3589,8 +3639,18 @@ pub struct FunctionType<'db> {
 
 #[salsa::tracked]
 impl<'db> FunctionType<'db> {
-    pub fn has_decorator(self, db: &dyn Db, decorator: Type<'_>) -> bool {
-        self.decorators(db).contains(&decorator)
+    pub fn has_known_class_decorator(self, db: &dyn Db, decorator: KnownClass) -> bool {
+        self.decorators(db).iter().any(|d| {
+            d.into_class_literal()
+                .is_some_and(|c| c.class.is_known(db, decorator))
+        })
+    }
```

Comment:
Would it make sense to have a bitflag on `FunctionType` where we initialize well known decorators?

Reply:
Maybe. I think this should wait until we have proper decorator support, at which point we won't track all decorators on `FunctionType` anymore at all. It may be that we still want to have a "shortcut" where we represent `classmethod(some_function)` as just a `FunctionType` with a flag on it, and similar for `staticmethod`. But that seems like the right time to optimize this representation, since all of this is temporary.